### PR TITLE
add VolumetricConvolution (3D convolution)

### PR DESCRIPTION
**Pull request by SeaOfOcean**
## What changes were proposed in this pull request?

add VolumetricConvolution (3D convolution)

## How was this patch tested?

unit test

### CODE REVIEW

**Comment by SeaOfOcean** [comment]
pr validate pass

**Comment by SeaOfOcean** [comment]
performance test pass

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricConvolution.scala [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+   * Computing the gradient of the module with respect to its own input. This is returned in
+   * gradInput. Also, the gradInput state variable is updated accordingly.
+   * @param input
+   * @param gradOutput
+   * @return
+   */
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    require(input.dim() == 4 || input.dim() == 5,
+      s"4D or 5D (batch mode) tensor expected for input, but got: ${ input.dim() }d")
+    require(input.isContiguous())
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricConvolution.scala]: Add error message.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricConvolution.scala [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+    require(gradOutput.isContiguous())
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricConvolution.scala]: Add error message.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricConvolution.scala [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+        }
+        gradBias.setValue(i + 1, ev.plus(gradBias.valueAt(i + 1),
+          ev.times(ev.fromType(scale), sum)))
+        i += 1
+      }
+    }
+  }
+
+  override def accGradParameters(input: Tensor[T], gradOutput: Tensor[T], scale: Double): Unit = {
+    require(input.isContiguous())
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricConvolution.scala]: Add error message.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricConvolution.scala [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+    require(gradOutput.isContiguous())
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricConvolution.scala]: Add error message.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricConvolution.scala [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+    output
+  }
+
+  private def updateGradInputFrame(gradInput: Tensor[T], gradOutput: Tensor[T], weight: Tensor[T],
+    fGradInput: Tensor[T], kT: Int, kW: Int, kH: Int, dT: Int, dW: Int, dH: Int,
+    pT: Int, pW: Int, pH: Int): Unit = {
+    val gradOutput2d = gradOutput.view(gradOutput.size(1),
+      gradOutput.size(2) * gradOutput.size(3) * gradOutput.size(4))
+    fGradInput.addmm(ev.fromType(0), fGradInput,
+      ev.fromType(1.0), weight, gradOutput2d)
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricConvolution.scala]: ev.one

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricConvolution.scala [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        NNPrimitive.unfoldedCopyVolDouble(fInput.asInstanceOf[Tensor[Double]],
+          input.asInstanceOf[Tensor[Double]], kT, kW, kH, dT, dW, dH, pT, pW, pH, nInputPlane,
+          inputDepth, inputWidth, inputHeight, outputDepth, outputWidth, outputHeight)
+      case FloatType =>
+        NNPrimitive.unfoldedCopyVolFloat(fInput.asInstanceOf[Tensor[Float]],
+          input.asInstanceOf[Tensor[Float]], kT, kW, kH, dT, dW, dH, pT, pW, pH, nInputPlane,
+          inputDepth, inputWidth, inputHeight, outputDepth, outputWidth, outputHeight)
+    }
+
+    output2d.addmm(ev.fromType[Int](0), output2d, ev.fromType[Int](1), weight, fInput)
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/VolumetricConvolution.scala]: ev.one ev.zero

**Pull request merged**